sm_ipt/doc/usage.rst

```diff
@@ -0,0 +1,228 @@
+
+.. code-block:: c
+
+	int send_nocopy(size_t len)
+	{
+		uint8_t *buffer;
+		size_t i;
+
+		sm_ipt_alloc_tx_buf(&sm_ipt_ctx, &buffer, len);
+
+		/* Use buffer as any other allocated buffer */
+		some_data_preparation_function(buffer);
+
+		return sm_ipt_send(&sm_ipt_ctx, buffer, len);
+	}
+
+The above code uses the :c:func:`sm_ipt_alloc_tx_buf` function to allocate the transmit buffer in the shared memory.
+The buffer is prepared by the :c:func:`some_data_preparation_function` and sent to another core through the :c:func:`sm_ipt_send` function.
+
+Data send abort
+---------------
+
+It is possible to free the transmit buffer instead of sending it.
+
+.. code-block:: c
+
+	int send_if(size_t len, bool send_choice)
+	{
+		uint8_t *buffer;
+		size_t i;
+
+		sm_ipt_alloc_tx_buf(&sm_ipt_ctx, &buffer, len);
+
+		if (send_choice){
+			/* Prepare some data */
+			some_data_preparation_function(buffer);
+
+			return sm_ipt_send(&sm_ipt_ctx, buffer, len);
+		}
+		else{
+			sm_ipt_free_tx_buf(&sm_ipt_ctx, buffer);
+
+			return ERROR_ABORT;
+		}
+	}
+
+The above code uses the :c:func:`sm_ipt_alloc_tx_buf` function to allocate the transmit buffer in the shared memory.
+The buffer is sent only if ``send_choice`` is true. If ``send_choice`` is false the buffer is free using the :c:func:`sm_ipt_free_tx_buf` and the send operation is aborted.
+
+Data receive full-copy
+----------------------
+
+When data is received, it can be processed using the full-copy mechanism.
+The buffer must be copied to the local memory.
+The shared memory rx buffer must be freed and data processing takes place on the local copy of the buffer.
+
+.. code-block:: c
+
+	volatile uint8_t receive_buffer[MAX_PACKET_SIZE];
+	volatile bool data_received;
+
+	void sm_ipt_receive_handler(const uint8_t *packet, size_t len)
+	{
+		/* Copy whole message */
+		memcpy(receive_buffer, packet, len);
+
+		/* Free receive buffer */
+		sm_ipt_free_rx_buf(&sm_ipt_ctx, packet);
+
+		/* Signal message received */
+		data_received = true;
+	}
+
+	void data_processing_thread(void)
+	{
+		/* Wait for incoming data */
+		while(!data_received);
+
+		/*Process data */
+		some_data_processing_function(receive_buffer);
+	}
+
+The above code uses the :c:func:`memcpy` function to copy the shared memory buffer into the local buffer.
+The next shared memory buffer is freed using the :c:func:`sm_ipt_free_rx_buf` function
+The data is processed locally using a copy of the buffer.
+
+Data receive no-copy
+--------------------
+
+When the data is received, it can be processed using the no-copy mechanism.
+The pointer to the shared memory buffer must be saved for later use.
+The shared memory rx buffer must be freed and data processing takes place on the shared memory buffer.
+
+.. note::
+   This approach increases the risk of out-of-memory errors if the buffer is not freed right after receiving data.
+
+.. code-block:: c
+
+	volatile void *receive_buffer;
+	volatile bool data_received;
```

Comment:
The same comment as above. Maybe something like that:

```c
/* Put the buffer pointer to some FIFO */
put_buffer_ptr_to_fifo(packet, len);
```

in `data_processing_thread`:

```c
/* Wait for incoming data */
get_buffer_ptr_from_fifo(&packet, &len);

/*Process data */
some_data_processing_function(receive_buffer);

/* Free receive buffer */
sm_ipt_free_rx_buf(&sm_ipt_ctx, packet);

```